### DISCUSSION TITLE
Laying out future AI

### DISCUSSION THREAD

**Discussion by RisingLeaf** [endless-sky/endless-sky#8173] [Balance, Mechanics, and Interactions]
Yes, we all know it, the current AI works. It also tends to do something called death-balling. This is not really obvious in small scale battles but even then, most of the time it is just two ships circulating around each other until they die.

So what I want to do in this discussion is the following:
- What features do we want for the AI so that we can get rid of death-balling. This is more the cc side of things
- How should these features of the AI be defined in data files, right now we only have personality which is really simple
- How will a new system interact with the old system, as we want to keep backward compatibility and dont want to break plugins

So, to get discussion going here is my idea:

## Combat roles
Combat roles are something you find in many games, some combat roles for example could be:
- tank
- fighter
- artillery(mid to long range)
but also:
- carrier (should not engage too much in combat as it can repair fighters)
- support (probably a bunch of roles such as repair or ships that pull other ships out of combat)

Combat roles would give us a lot more room to adjust ship behavior and would let us integrate things we do not have yet (e.g. support ships). These roles should probably be able to be defined ship specific and not only for the whole fleet.
Also, the player should be able to assign combat roles to their ships. Maybe we could even create a combat mode that triggers the combat roles of the players ship and otherwise they will behave as usual.

Here is an example of how I would see this implemented in data:
```
fleet "Navy Interception Fleet"
    personality heroic
        confusion 10
    variant
        "Cruiser"
            role tank
        "Aerie"
            role artillery
```

Included in this would also be to auto assign roles to ships in fleets:
The idea is to look what loadout a ship has and depending on this give the ship a combat role, if none is already defined.

Obviously we have a long for things like this to get in and it will take a lot of PRs. But we have to start somewhere, right?
So please give your opinions or ideas.

**Comment by Hurleveur** [comment]
https://github.com/endless-sky/endless-sky/pull/7552 I already did the artillery AI and the framework for more AIs in this PR

**Reply by RisingLeaf** [reply]
I think this one is a good base, yes

**Comment by Zitchas** [comment]
I like the idea of being able to designate ships as being certain roles. I like it even more if I, as the player, can manually specify it for ships in my fleet. 

Sort of a companion bit to this: Being able to specify that certain weapons and/or hardpoints are "defensive only," and then having the AI ignore defensive weapons when picking ranges, could help with this. For instance, if a capital ship has some lightweight turrets designed for shooting fighters, and a bunch of big heavy long range turrets for slugging with other ships; it could be instructed to ignore the anti-fighter turrets and pick ranges suited to the heavier ones. If ships just naturally skip considering defensive weapons, then they would organically start to sort themselves out into longer ranges engagements.

**Reply by Hurleveur** [reply]
whilst I agree with that that's only on the player's side of things, not on AI
more customization always is good though, especially for personalities

**Reply by Zitchas** [reply]
Well, if the hardpoints can be labeled as "defensive" than that applies to the AI too. We just specify it right in the ship definition.

**Reply by RisingLeaf** [reply]
I don't think something like this should be done in the ship definition but rather somewhere in the fleet definition.
The player would need some sort of tactical panel where stuff like this could be done.

**Reply by Zitchas** [reply]
Well, as an example, Vitalchip tried adding extra turrets to a variety of ships; but designated them as defensive. They also changed the code so that defensive turrets are basically *always* opportunistic and autofire, and ignore the target the ship is actually locked onto. That meant that they could effectively increase the number of turrets on a ship and provide hardpoints for AM and anti-fighter defense, but without significantly altering the focused offensive power that the ship can bring to bear. 

I'm not saying it's a perfect solution, but it is an option that allows for ships to be given hardpoints for an AM without also giving them the ability to add additional focused offensive power. Offensive hardpoints could be designated defensive by the player, but defensive ones would be locked to prevent them unbalancing offensive capabilities further.

Here's a working example:
Off = Offensive (player/AI controlled weapon
Def = Defensive (locked to opportunistic autofire weapon)
AM= Antimissile

Yellow/orange lines designate gunports
Blue designate turrets
Green designate pylons (hardpoints that can only be used for missiles)

The numbers are the range of the weapon.

![Weapons Control Panel](https://user-images.githubusercontent.com/32169904/213940845-06d2c260-adaa-42ca-954f-76281eb6dbab.png)

**Comment by RisingLeaf** [comment]
Big Update:
Hurls PR for expanding the artillery AI just got merged, this means we have the groundwork done, from this point on additional AI expansion should be a lot easier.

**Reply by Hurleveur** [reply]
This adds a ramming AI personality that just rushes in and (automatically set) artillery, that runs away, and non suicidary AI with heavy rockets.
We could make an AI node to define them, and add new ones, for instance defining safe range in it, and whatever comes next with AI improvements

**Comment by RisingLeaf** [comment]
I think we maybe want to get away from personalities for combat, what I think could work is having an ai node that lets you define one combat behaviour and then you should be able to assign this behaviour to ships of a fleet

**Reply by Zitchas** [reply]
I think there's a place for both.

Personalities should direct a ship's preferences, trends, and overrides. But most ships should be light on that and mostly governed by logic of what they have available.

For example, a reckless combatant might have a c "brawler" personality that pushes it to agressively close and fight at the closest possible distance, no matter what weapons it has. But the default AI should look at the weapon and decide that, since the current target is far away and the AI has mostly long ranged weaponry, it should opt to stay as far as possible away, even if it means it isn't using *all* its weaponry.

**Comment by RisingLeaf** [comment]
So here is what I am going to do:
I will mess with the code a bit to get the minimum viable product of what I want, I will open a draft PR for people to test it.
I will link it here as soon as it is finished.

I think that way we can get a feeling of how the mechanic should look like.

**Reply by movingpictures** [reply]
> So here is what I am going to do: I will mess with the code a bit to get the minimum viable product of what I want, I will open a draft PR for people to test it. I will link it here as soon as it is finished.
> 
> I think that way we can get a feeling of how the mechanic should look like.

Will be glad to test this when it's done. Replying to get updates. 

Gracias.

**Comment by RisingLeaf** [comment]
Because we are currently in a development phase of wrapping stuff up and fixing bugs, I will postpone this until 0.10.0 is out.
